spikeinterface/sortingcomponents/motion_estimation.py

```diff
@@ -793,14 +865,136 @@ def jac(p):
+                    format="csr",
+                )
+                block_disp_pairs = np.concatenate(
+                    (block_disp_pairs, np.zeros(T - 1)),
+                )
 
-    return displacement
+            coefficients.append(block_sparse_kron)
+            targets.append(block_disp_pairs)
+        coefficients = sparse.block_diag(coefficients)
+        targets = np.concatenate(targets, axis=0)
+
+        # spatial smoothness prior: penalize difference of each block's
+        # displacement with the next.
+        # only if B > 1, and not in the last window.
+        # this is a (BT, BT) sparse matrix D such that:
+        # entry at (i, j) is:
+        #  {   1 if i = j, i.e., i = j = bT + t for b = 0,...,B-2
+        #  {  -1 if i = bT + t and j = (b+1)T + t for b = 0,...,B-2
+        #  {   0 otherwise.
+        # put more simply, the first (B-1)T diagonal entries are 1,
+        # and entries (i, j) such that i = j - T are -1.
+        if B > 1 and spatial_prior:
+            spatial_diff_operator = sparse.diags(
+                (
+                    np.ones((B - 1) * T, dtype=block_sparse_kron.dtype),
+                    np.full((B - 1) * T, -1, dtype=block_sparse_kron.dtype),
+                ),
+                offsets=(0, T),
+                shape=((B - 1) * T, B * T),
+            )
+            coefficients = sparse.vstack((coefficients, spatial_diff_operator))
+            targets = np.concatenate((targets, np.zeros((B - 1) * T, dtype=targets.dtype)))
+
+        # initialize at the column mean of pairwise displacements (in each window)
+        p0 = D.mean(axis=2).reshape(B * T)
+
+        # use LSMR to solve the whole problem
+        displacement, *_ = sparse.linalg.lsmr(coefficients, targets, x0=p0)
+        displacement = displacement.reshape(B, T).T
+
+        # TODO: do we need to weight the upsampling somehow?
+    else:
+        raise ValueError(f"Method {convergence_method} doesn't exist for compute_global_displacement")
+
+    # try to avoid spurious constant offsets
+    # let the user choose how to do this. here are some ideas.
+    # (one can also -= their own number on the result of this function.)
+    if reference_displacement == "mean":
+        displacement -= displacement.mean()
+    elif reference_displacement == "median":
+        displacement -= np.median(displacement)
+    elif reference_displacement == "mode_search":
+        # just a sketch of an idea -- things might want to change.
+        step_size = 0.1
+        round_mode = np.round  # floor?
+        best_ref = np.median(displacement)
+        max_zeros = np.sum(round_mode(displacement - best_ref) == 0)
+        for ref in np.arange(np.floor(displacement.min()), np.ceil(displacement.max()), step_size):
+            n_zeros = np.sum(round_mode(displacement - ref) == 0)
+            if n_zeros > max_zeros:
+                max_zeros = n_zeros
+                best_ref = ref
+        displacement -= best_ref
```

Comment:
From what I understand the continuity of spatial dim is handle already by spatial_prior.
Here we are handeling the "common" stuff and not by windowd stuff.
Am I correct ? If so I would vote for median by default.
Empirical mode sometimes fails due to the bin size. No ?
Did you ebnchmark this "mode_search" ?

Reply:
Mode search is a little silly -- could be sped up with a call to histogram anyway. I'd be happy with median by default.